### PR TITLE
adding container registry webhook module

### PR DESCRIPTION
**Pull request by zikalino**
##### SUMMARY
Completing container registry scenario. Adding webhook module

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
azure_rm_containerregistrywebhook

##### ANSIBLE VERSION
2.7

### CODE REVIEW

File: test/integration/targets/azure_rm_containerregistry/tasks/main.yml [hunk 1]
```diff
@@ -1,118 +1,214 @@
...
+      - output.sku == 'Standard'
+      - output.tags['NewTag'] == 'newtag'
+      - output.credentials | length == 0
+      - output.credentials['password'] is not defined
+      - output.credentials['password2'] is not defined
+
+- name: Gather facts for single Container Registry
+  azure_rm_containerregistry_facts:
+    resource_group: "{{ resource_group }}"
+    name: "acr{{ rpfx }}"
```
Comment (Fred-sun) [inline test/integration/targets/azure_rm_containerregistry/tasks/main.yml:71]: azure_rm_containerregistry_facts module's parameter is "registry_name", not name.
```e.g.
fatal: [localhost]: FAILED! => {"changed": false, "msg": "Unsupported parameters for (azure_rm_containerregistry_facts) module: name Supported parameters include: ad_user, adfs_authority_url, api_profile, append_tags, auth_source, cert_validation_mode, client_id, cloud_environment, password, profile, registry_name, resource_group, secret, subscription_id, tags, tenant"}
	to retry, use: --limit @/home/fred/task/45394/45394.retry

File: test/integration/targets/azure_rm_containerregistry/tasks/main.yml [hunk 2]
```diff
@@ -1,118 +1,214 @@
...
+    that:
+      - output.changed == False
+      - output.registries[0]['name'] != None
+      - output.registries[0]['location'] != None
+      - output.registries[0]['admin_user_enabled'] != None
+      - output.registries[0]['sku'] != None
+      - output.registries[0]['provisioning_state'] != None
+      - output.registries[0]['login_server'] != None
+      - output.registries[0]['id'] != None
+      - output.registries[0]['credentials'] != None
```
Comment (Fred-sun) [inline test/integration/targets/azure_rm_containerregistry/tasks/main.yml:85]: azure_rm_containerregistry_facts output is a dictionary, not a list. there is no credentials variable in output. just like below.
```e.g.
ok: [localhost] => {
    "output": {
        "ansible_facts": {}, 
        "changed": false, 
        "failed": false, 
        "registries": {
            "acrtest45394": {
                "admin_user_enabled": false, 
                "creation_date": "2018-09-26T08:57:47.322029Z", 
                "id": "/subscriptions/2085065b-00f8-4cba-9675-ba15f4d4ab66/resourceGroups/myRG02/providers/Microsoft.ContainerRegistry/registries/acrtest45394", 
                "location": "eastus2", 
                "login_server": "acrtest45394.azurecr.io", 
                "name": "acrtest45394", 
                "provisioning_state": "Succeeded", 
                "sku": {
                    "name": "Standard", 
                    "tier": "Standard"
                }, 
                "tags": {
                    "Environment": "Production", 
                    "NewTag": "newtag", 
                    "Release": "beta1"
                }, 
                "type": "Microsoft.ContainerRegistry/registries"
            }
        }, 
        "warnings": [
            "Azure API profile latest does not define an entry for ContainerRegistryManagementClient"
        ]
    }
}

File: test/integration/targets/azure_rm_containerregistry/tasks/main.yml [hunk 3]
```diff
@@ -1,118 +1,214 @@
...
+    name: "acr{{ rpfx }}"
+  register: output 
+
+- name: Assert that facts are returned
+  assert:
+    that:
+      - output.changed == False
+      - output.registries[0]['name'] != None
+      - output.registries[0]['location'] != None
+      - output.registries[0]['admin_user_enabled'] != None
```
Comment (Fred-sun) [inline test/integration/targets/azure_rm_containerregistry/tasks/main.yml:80]: The containerregistry's 'admin_user_enabled' is false when created or update. it need change "- output.registries[0]['admin_user_enabled'] != None" to " - output.registries[0]['admin_user_enabled'] == False"

**Comment by Fred-sun** [comment]
@zikalino

**Comment by Fred-sun** [comment]
@zikalino  Could you help recheck this PR? Thanks!

**Comment by Fred-sun** [comment]
@zikalino

**Comment by Fred-sun** [comment]
@zikalino Could you help recheck this PR and move this PR to forward? Thanks!

File: lib/ansible/modules/cloud/azure/azure_rm_containerregistrywebhook.py [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+---
+module: azure_rm_containerregistrywebhook
+version_added: "2.8"
+short_description: Manage Webhook instance.
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_containerregistrywebhook.py:20]: Azure Container Registry

File: lib/ansible/modules/cloud/azure/azure_rm_containerregistrywebhook.py [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+            - 'enabled'
+            - 'disabled'
+    scope:
+        description:
+            - "The scope of repositories where the event can be triggered. For example, 'foo:*' means events for all tags under repository 'foo'. 'foo:bar' m
+              eans events for 'foo:bar' only. 'foo' is equivalent to 'foo:latest'. Empty means all events."
+    actions:
+        description:
+            - The list of actions that trigger the webhook to post notifications.
+        type: list
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_containerregistrywebhook.py:59]: choices of list element?

File: lib/ansible/modules/cloud/azure/azure_rm_containerregistrywebhook.py [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+                                              supports_tags=False)
+
+    def exec_module(self, **kwargs):
+        """Main module execution method"""
+
+        for key in list(self.module_arg_spec.keys()):
+            if hasattr(self, key):
+                setattr(self, key, kwargs[key])
+            elif kwargs[key] is not None:
+                if key == "location":
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_containerregistrywebhook.py:182]: refine below codes?

**Comment by Fred-sun** [comment]
@zikalino Could you help recheck the PR and update according yungezz's comment? Thanks!

**Comment by Fred-sun** [comment]
@zikalino

**Comment by Fred-sun** [comment]
@zikalino

**Comment by Fred-sun** [comment]
@zikalino could you help recheck this PR according by yungezz's comment? Thanks!

**Comment by Fred-sun** [comment]
@zikalino

**Comment by Fred-sun** [comment]
@zikalino

**Comment by Fred-sun** [comment]
kindly ping

**Comment by Fred-sun** [comment]
@zikalino Please help take a look this PR when you're available! Thanks!

**Comment by Fred-sun** [comment]
@zikalino Please update this PR when you're available! Thanks!

**Comment by Fred-sun** [comment]
@zikalino

**Comment by Fred-sun** [comment]
@zikalino Please help resolve the conflicting files, resolve the comment and push for review! Thank you very much!

**Comment by Fred-sun** [comment]
@zikalino Please resolve the conflicting file and changed according by comments! Thank you very much!

**Comment by Fred-sun** [comment]
@zikalino When you are free, please help to solve the PR mistake. Thank you very much!

**Comment by Fred-sun** [comment]
@zikalino Need you to update! Please update when you're free! Thank you very much!

**Comment by Fred-sun** [comment]
@MyronFanQiu The azure_rm_containerregistrywebhook module has been added to ansible 2.9.0devel. Can we close the PR? Please confirm! Thank you very much!

**Pull request closed without merging**